app/src/main/java/com/github/epfl/meili/map/MapActivity.kt

```diff
@@ -58,20 +71,86 @@ class MapActivity : AppCompatActivity(), OnMapReadyCallback, GoogleMap.OnPoiClic
         mapFragment?.getMapAsync(this)
     }
 
+    private fun setUpClusterer() {
+        if (isPermissionGranted()) {
+            val locationService = LocationService()
+            locationService.listenToLocationChanges(poiMarkerViewModel)
+        }
+
+
+        poiMarkerViewModel.setPoiService(PoiService())
+
+        val currentUser = Auth.getCurrentUser()
+        if (currentUser != null) {
+            poiMarkerViewModel.setDatabase(FirestoreDatabase("users-poi-list/${currentUser.uid}/poi-list", PointOfInterest::class.java))
+        }
+
+
+        // Initialize the manager with the context and the map.
+        // (Activity extends context, so we can pass 'this' in the constructor.)
+        clusterManager = ClusterManager(this, map)
+
+        clusterRenderer = PoiRenderer(this, map, clusterManager)
+
+        clusterManager.renderer = clusterRenderer
+
+        // Point the map's listeners at the listeners implemented by the cluster
+        // manager.
+        map.setOnCameraIdleListener(clusterManager)
+        map.setOnMarkerClickListener(clusterManager)
+
+        // Add on click listener
+        clusterManager.setOnClusterItemClickListener {
+            val intent = Intent(this, PoiActivity::class.java)
+            intent.putExtra(POI_KEY, it.poi)
+
+            if(poiMarkerViewModel.mPointsOfInterestStatus.value?.get(it.poi) ==PoiMarkerViewModel.PointOfInterestStatus.REACHABLE){
+                poiMarkerViewModel.setPoiVisited(it.poi)
+            }
+
+            startActivity(intent)
+            true
+        }
+
+        poiMarkerViewModel.mPointsOfInterestStatus.observe(this) {
+            addItems(it)
+        }
+    }
+
+    private fun addItems(map: Map<String, PoiMarkerViewModel.PointOfInterestStatus>) {
+        val newMap = HashMap<PoiItem, PoiMarkerViewModel.PointOfInterestStatus>()
+        for (entry in map.entries) {
+            val poiItem: PoiItem
+            if (poiItems.containsKey(entry.key)) {
```

Comment:
You could use the getOrDefault method of map I think

Reply:
Unfortunately it's not possible with this APK version, because poiItems is a Map instance and not HashMap

Reply:
Cool :+1: